router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -55,31 +117,41 @@ class StreamStatsFilter(statsReceiver: StatsReceiver)
       statsReceiver.scope("stream"),
       durationName = Some("total_latency")
     )
-  private[this] val rspFutureStats =
-    new StreamStats(
-      statsReceiver,
-      durationName = Some("request_latency"),
-      successName = Some("")
-    )
 
-  override def apply(req: Request, service: Service[Request, Response]): Future[Response] = {
+  override def apply(req0: Request, service: Service[Request, Response]): Future[Response] = {
     reqCount.incr()
     val reqT = Stopwatch.start()
+    val req1 = Request(req0.headers, reqStreamStats(reqT)(req0.stream))
 
-    req.stream.onEnd.respond(reqStreamStats(reqT))
+    @inline def classify(rsp: Try[Response])(frame: Option[Try[Frame]]): Unit =
+      classifier(H2ReqRep(req1, rsp.map((_, frame)))) match {
+        case Successful(_) => successes.incr()
+        case _ => failures.incr()
+      }
 
-    service(req)
+    service(req1)
+      .transform {
+        case Return(response) =>
+          val rspT = Stopwatch.start()
+          val stream = rspStreamStats(rspT, classify(Return(response)))(response.stream)
```

Comment:
I'd explicitly name onFinalFrame since it's sort of an optional parameter:

```
rspStreamStats(rspT, onFinalFrame = classify(Return(response)))
```

This helps clarify that `reqStreamStats` is the same thing but doesn't provide any `onFinalFrame` callback